### DISCUSSION TITLE
i need help to convert this cleaner from javascript to java

### DISCUSSION THREAD

**Discussion by vidhun73** [clojurephant/clojurephant#194] [General]
Cleaner is usually some object (helper). Which collect actions what should be done at the end of a test. For example if we create a db table in scope of the test then we should immediately inform cleaner that we need delete the table.

Cleaner is executed at the and of the test (part "after").
It is executed always if test passed or failed.
All tasks in cleaner are stored in stack LIFO.
If cleaner task fail then we need log it but continue with cleaning (try catch statement should be there probably).
Michal did not found any cleaner in TestNG. We should probably implement it. Could be added into context. If we use annotation @BeforeTest for creation and adding into context and @AfterTest for execution then it should be enough. Source: https://howtodoinjava.com/testng/testng-before-and-after-annotations/

Attaching cleaner.js what we have in our javascript solution. Could be refactored to java.

class Cleaner {
    constructor(my_logger) {
        this.cleaning_tasks_to_be_executed = [];
        this.enable();
        this.my_logger = my_logger;
    }

    async enable() {
        this.is_enable = true;
    };
    async disable() {
        this.is_enable = false;
    };

    // Only sync functions could be added
    async add(func) {
        let cleaning_task = {
            function: func,
            should_be_executed: this.is_enable
        }
        this.cleaning_tasks_to_be_executed.push(cleaning_task);
    };

    async clean() {
        let clean_after_started_miliseconds = new Date().valueOf();

        let task = 1;
        while (this.cleaning_tasks_to_be_executed.length) {
            let cleaning_task = this.cleaning_tasks_to_be_executed.pop();

            if(!cleaning_task.should_be_executed) {
                this.my_logger.info("Cleaner: Skipping - Function: " + cleaning_task.function);
                continue;
            }

            let task_after_started_miliseconds = new Date().valueOf();
            
            try {
                this.my_logger.info("Cleaner: Running - Function: " + cleaning_task.function);
                await cleaning_task.function();
                this.my_logger.info("Cleaner: Done - Function: " + cleaning_task.function);
            }
            catch (err) {
                this.my_logger.info("Cleaner: Failed - Function: " + cleaning_task.function + "\nError: " + err);
            }
            
            let task_after_done_miliseconds = new Date().valueOf() - task_after_started_miliseconds;
            this.my_logger.info(`Cleaning task ${task} - done in: ${task_after_done_miliseconds}ms`);
            task++;
        }
        
        let clean_after_done_miliseconds = new Date().valueOf() - clean_after_started_miliseconds;
        this.my_logger.info("Clean - done in: " + clean_after_done_miliseconds + "ms");
    }
}

module.exports = Cleaner;

**Comment by ajoberstar** [comment]
@vidhun73 This repository is for a Clojure/ClojureScript Gradle plugin, so I think you'll want to ask this JS to Java question elsewhere.
